src/prefect/engine/result/gcs_result.py

```diff
@@ -0,0 +1,137 @@
+import base64
+from typing import TYPE_CHECKING, Any, Optional
+
+import cloudpickle
+
+from prefect.engine.result.base import Result
+from prefect.client import Secret
+from prefect.utilities import logging
+
+if TYPE_CHECKING:
+    import google.cloud
+
+
+class GCSResult(Result):
+    """
+    Result that is written to and read from a Google Cloud Bucket.
+
+    To authenticate with Google Cloud, you need to ensure that your flow's
+    runtime environment has the proper credentials available
+    (see https://cloud.google.com/docs/authentication/production for all the authentication options).
+
+    You can also optionally provide the name of a Prefect Secret containing your
+    service account key.
+
+    Args:
+        - value (Any, optional): the value of the result
+        - bucket (str): the name of the bucket to write to / read from
+        - credentials_secret (str, optional): the name of the Prefect Secret
+            which stores a JSON representation of your Google Cloud credentials.
+        - **kwargs (Any, optional): any prefect.engine.result.Result options
+    """
+
+    def __init__(
+        self,
+        value: Any = None,
+        bucket: str = None,
+        credentials_secret: str = None,
+        **kwargs: Any
+    ) -> None:
+        self.bucket = bucket
+        self.credentials_secret = credentials_secret
+        self.logger = logging.get_logger(type(self).__name__)
+        super().__init__(value, **kwargs)
+
+    @property
+    def gcs_bucket(self) -> "google.cloud.storage.bucket.Bucket":
+        if not hasattr(self, "_gcs_bucket"):
+            from prefect.utilities.gcp import get_storage_client
+
+            if self.credentials_secret:
+                credentials = Secret(self.credentials_secret).get()
+            else:
+                credentials = None
+            client = get_storage_client(credentials=credentials)
+            self.gcs_bucket = client.bucket(self.bucket)
+        return self._gcs_bucket
+
+    @gcs_bucket.setter
+    def gcs_bucket(self, val: Any) -> None:
+        self._gcs_bucket = val
+
+    def __getstate__(self) -> dict:
+        state = self.__dict__.copy()
+        if "_gcs_bucket" in state:
+            del state["_gcs_bucket"]
+        return state
```

Comment:
Why remove the `_gcs_bucket`? jw for my own knowledge

Reply:
The `gcs_client.bucket` is not serializable, this removes the offending object from state that is used for pickling.